### DISCUSSION TITLE
Can we present .mm files on blog or website?

### DISCUSSION THREAD

**Discussion by sgalcheung** [freeplane/freeplane#1876] [Ideas]
I found Xmind implement this [xmind-embed-viewer](https://github.com/xmindltd/xmind-embed-viewer)
![image](https://github.com/freeplane/freeplane/assets/42566429/061b5151-1da9-46a9-bb5a-07ea832f202f)

So I have this idea.

**Comment by abitrolly** [comment]
If .mm files can be transformed to the format that https://github.com/drichard/mindmaps understands, then it is possible to patch it to generate static mindmap sites.

**Reply by abitrolly** [reply]
There are actually much more variants - https://github.com/topics/mindmap

`markmap` is particularly interesting https://github.com/markmap/markmap

**Comment by sgalcheung** [comment]
I think the first thing is to transform .mm files because the freeplane depends on `org.apache.xmlgraphics`.

**Reply by abitrolly** [reply]
From https://github.com/freeplane/freeplane/blob/1.12.x/freeplane/doc/freeplaneFunctions.mm and https://docs.freeplane.org/attic/old-mediawiki-content/Home.html?highlight=functions#functions it doesn't seem complicated.

I guess it is just traversing from root to child `<node>` objects to get relations and extracting `TEXT` attribute from them.

Should be 10 minutes kata, but I am lazy. )

**Reply by sgalcheung** [reply]
Can we resolve it in Node.js, like TypeScript, which also means `org.apache.xmlgraphics` has an alternative？

Do we use another language to implement this logic again?

**Reply by abitrolly** [reply]
Reading XML works in whatever language you want. I would use Python and start with https://github.com/romilly/fm2md which already does conversion to markdown. Haven't tested it though.

**Comment by dartrax** [comment]
My favorite would be to save as .SVG and embed the .mm-format inside the SVG, like draw.io does with "*.drawio.svg" files.
Edit: I forgot this link: https://github.com/freeplane/freeplane/issues/1151

However, until this happens (if ever):
freeplane can export as Java Applet, and you can use cheerpj to display in the browser. Works, but takes some time to load.

**Comment by sgalcheung** [comment]
Oh, I noticed that Freemind can export as .svg file, but this still meant using Freemind to convert a .mm file to.svg file, I say above that it is export function dependent `org.apache.xmlgraphics`.

**Reply by dartrax** [reply]
If you want to do it without converting, you'd need to go the Java Applet path. You can upload the Applet to the server, configure using with cheerpj, and then just provide the .mm to display it on a website. You can try it here (Loading may need a minute):
https://www.dartrax.de/other/freeplane/map_cheerpj.html

**Reply by sgalcheung** [reply]
Do we have a chance to split this dependent, even not dependent on a specific language?
Or use full Node.js to implement it?

**Reply by abitrolly** [reply]
@sgalcheung sounds like you want .mm files to be .html files. And it still will be "save" instead of "export" to produce them, and it still would need to be uploaded to the website. XML .mm format is pretty independent from specific language.

**Reply by dpolivaev** [reply]
It looks you do not understand what kind of data is contained in the mm files, and which role plays `org.apache.xmlgraphics`. Let me explain. The mm file contains data about the mind map. `org.apache.xmlgraphics` is a library used for generating svg and pdf exports and for drawing svg images inside mind maps. The generated svg files have no links and no animations like folding and unfolding, they are just vector graphics.

Freeplane itself contains no functionality for displaying mind maps in web sites since Java Applet technology was removed from modern browsers and Java versions. One could develop new not Java based software for displaying mind maps, and there are a couple of such applications. However, because of the huge feature set  on Freeplane, I do not think that any complete solution can be created. And creating such app is not in my focus. A similar situation relates to mobile versions that should also be written from scratch because libraries used by Freeplane are available only for desktop computers.

By the way, I appreciate if people do not mix up FreeMind and Freeplane.

Regards.

**Reply by sgalcheung** [reply]
@dpolivaev 
Thank you for the tip, I don't want to edit .mm file in the browser, first I want to preview it in the browser website(blog or any others). although this project: [kityminder](https://github.com/fex-team/kityminder/blob/dev/src/protocol/freemind.js) support Freemind, but this algorithm is not correct, it forgets the correct position of the mind map node.

**Reply by beccon4** [reply]
There is a Nextcloud App based on Kityminder which is able to open mm Files - change them but not save them :-) It is abandonware somehow.

https://apps.nextcloud.com/apps/files_mindmap

**Reply by sgalcheung** [reply]
@beccon4 Not relevant apps, like Nextcloud, this discussion talks about from Freeplane to synchronize the same algorithm.
